datago/src/generator_http.rs

```diff
@@ -0,0 +1,403 @@
+use reqwest::header::HeaderMap;
+use reqwest::header::HeaderValue;
+use reqwest::header::AUTHORIZATION;
+use reqwest::Url;
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct SourceDBConfig {
+    pub sources: String,
+
+    #[serde(default)]
+    pub sources_ne: String,
+
+    #[serde(default)]
+    pub require_images: bool,
+
+    #[serde(default)]
+    pub require_embeddings: bool,
+
+    #[serde(default)]
+    pub tags: String,
+
+    #[serde(default)]
+    pub tags_ne: String,
+
+    #[serde(default)]
+    pub has_attributes: String,
+
+    #[serde(default)]
+    pub lacks_attributes: String,
+
+    #[serde(default)]
+    pub has_masks: String,
+
+    #[serde(default)]
+    pub lacks_masks: String,
+
+    #[serde(default)]
+    pub has_latents: String,
+
+    #[serde(default)]
+    pub lacks_latents: String,
+
+    #[serde(default)]
+    pub min_short_edge: i32,
+
+    #[serde(default)]
+    pub max_short_edge: i32,
+
+    #[serde(default)]
+    pub min_pixel_count: i32,
+
+    #[serde(default)]
+    pub max_pixel_count: i32,
+
+    #[serde(default)]
+    pub duplicate_state: i32,
+
+    #[serde(default)]
+    pub random_sampling: bool,
+    pub page_size: usize,
+}
+
+// TODO: Derive from the above
+#[derive(Debug, Serialize, Deserialize)]
+pub struct DbRequest {
+    pub fields: String,
+    pub sources: String,
+    pub sources_ne: String,
+    pub page_size: String,
+
+    pub tags: String,
+    pub tags_ne: String,
+
+    pub has_attributes: String,
+    pub lacks_attributes: String,
+
+    pub has_masks: String,
+    pub lacks_masks: String,
+
+    pub has_latents: String,
+    pub lacks_latents: String,
+    pub return_latents: String,
+
+    pub min_short_edge: String,
+    pub max_short_edge: String,
+
+    pub min_pixel_count: String,
+    pub max_pixel_count: String,
+
+    pub duplicate_state: String,
+    pub random_sampling: bool,
+
+    pub partitions_count: String,
+    pub partition: String,
+}
+
+// implement a helper to get the http request which correspons to the db request structure above
+impl DbRequest {
+    pub fn get_http_request(&self, api_url: &str, api_key: &str) -> reqwest::blocking::Request {
+        let mut url = if self.random_sampling {
+            Url::parse(&format!("{}images/random/", api_url)).unwrap()
+        } else {
+            Url::parse(&format!("{}images/", api_url)).unwrap()
+        };
+
+        let mut query_pairs = url.query_pairs_mut();
+
+        let maybe_add_field =
+            |query_pairs: &mut url::form_urlencoded::Serializer<url::UrlQuery>,
+             field: &str,
+             value: &str| {
+                if !value.is_empty() {
+                    query_pairs.append_pair(field, value);
+                }
+            };
```

Comment:
nit: You can take a mutable borrow of query_paris inside this closure, to avoid repeating it

```suggestion
        let query_pairs = &mut query_pairs;
        let maybe_add_field =
            move |field: &str,
             value: &str| {
                if !value.is_empty() {
                    query_pairs.append_pair(field, value);
                }
            };
```

Reply:
I don´t get the variable shadowing here, it's on purpose ? looks like dark magic to me, it's a rust thing ?

Reply:
seems to work, this syntax is really strange but ok, learnt something